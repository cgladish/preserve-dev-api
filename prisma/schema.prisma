// This is your Prisma schema file,
// learn more about it in the docs: https://pris.ly/d/prisma-schema

generator client {
  provider = "prisma-client-js"
}

datasource db {
  provider = "postgresql"
  url      = env("DATABASE_URL")
}

model User {
  id        Int       @id @default(autoincrement())
  email     String    @unique
  username  String    @db.VarChar(255)
  snippets  Snippet[]
  createdAt DateTime  @default(now())
  updatedAt DateTime  @updatedAt

  @@index([email, username])
}

model App {
  id        Int       @id @default(autoincrement())
  name      String    @db.VarChar(255)
  snippets  Snippet[]
  createdAt DateTime  @default(now())
  updatedAt DateTime  @updatedAt
}

model Snippet {
  id                  Int       @id @default(autoincrement())
  public              Boolean   @default(false)
  title               String?   @db.VarChar(255)
  appSpecificDataJson String?   @db.VarChar(255)
  views               Int
  creator             User      @relation(fields: [creatorId], references: [id])
  creatorId           Int
  App                 App       @relation(fields: [appId], references: [id])
  appId               Int
  messages            Message[]
  createdAt           DateTime  @default(now())
  updatedAt           DateTime  @updatedAt
}

model Message {
  id                  Int      @id @default(autoincrement())
  content             String   @db.VarChar(4096)
  sentAt              DateTime
  appSpecificDataJson String?  @db.VarChar(255)
  authorUsername      String   @db.VarChar(255)
  authorIdentifier    String?  @db.VarChar(255)
  authorAvatarUrl     String?  @db.VarChar(255)
  Snippet             Snippet  @relation(fields: [snippetId], references: [id])
  snippetId           Int
}
