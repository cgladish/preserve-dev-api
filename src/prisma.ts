import { PrismaClient } from "@prisma/client";
import Hashids from "hashids";

const PADDED_ID_LENGTH = 7;

type IdUtils = {
  idToExternalId: (id: number) => string;
  externalIdToId: (externalId: string) => number;
};
export const makeIdUtils = (modelName: string): IdUtils => ({
  idToExternalId: (id: number): string =>
    new Hashids(modelName, PADDED_ID_LENGTH).encode(id),
  externalIdToId: (externalId: string): number => {
    const id = Number(
      new Hashids(modelName, PADDED_ID_LENGTH).decode(externalId)[0]
    );
    if (Number.isNaN(id)) {
      throw new Error("Invalid external ID");
    }
    return id;
  },
});

const prisma = new PrismaClient();
const modelNames: (keyof typeof prisma)[] = [
  "app",
  "comment",
  "message",
  "snippet",
  "user",
];
modelNames.forEach((modelName) =>
  Object.assign(prisma[modelName], makeIdUtils(modelName))
);
export type ExtendedPrismaClient = PrismaClient & {
  app: PrismaClient["app"] & IdUtils;
  comment: PrismaClient["comment"] & IdUtils;
  message: PrismaClient["message"] & IdUtils;
  snippet: PrismaClient["snippet"] & IdUtils;
  user: PrismaClient["user"] & IdUtils;
};
export default prisma as ExtendedPrismaClient;
