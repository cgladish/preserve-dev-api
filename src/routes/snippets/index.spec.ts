import app from "../../index";
import prisma from "../../prisma";
import request from "supertest";
import { CreateSnippetInput } from ".";
import { set } from "lodash";
import { randText } from "@ngneat/falso";

describe("snippets routes", () => {
  describe("GET /:id", () => {
    it("can get a snippet by external ID", async () => {
      const appEntity = await prisma.app.create({
        data: {
          name: "Discord",
        },
      });
      const creatorEntity = await prisma.user.create({
        data: {
          username: "crasken",
          email: "chase.gladish@gmail.com",
        },
      });
      const snippetEntity = await prisma.snippet.create({
        data: {
          appId: appEntity.id,
          public: true,
          title: "Test snippet title",
          appSpecificDataJson: '{"key":"value"}',
          creatorId: creatorEntity.id,
          messages: {
            create: [
              {
                content: "Content",
                sentAt: new Date(10).toISOString(),
                appSpecificDataJson: '{"key2":"value2"}',
                authorUsername: "Icyspawn",
                authorIdentifier: "1234",
                authorAvatarUrl: "http://example.com/123.png",
              },
            ],
          },
        },
      });

      const response = await request(app)
        .get(`/snippets/${prisma.snippet.idToExternalId(snippetEntity.id)}`)
        .expect(200);
      expect(response.body).toMatchSnapshot();
    });

    it("returns 404 if no entity found", async () => {
      await request(app)
        .get(`/snippets/${prisma.snippet.idToExternalId(1)}`)
        .expect(404);
    });
  });

  describe("POST /", () => {
    it.each([
      ["appId", null],
      ["public", null],
      ["messages", null],
      ["messages[0].content", null],
      ["messages[0].sentAt", null],
      ["messages[0].authorUsername", null],
      ["appId", ""],
      ["messages[0].content", ""],
      ["messages[0].authorUsername", ""],
      ["appId", randText({ charCount: 21 })],
      ["title", randText({ charCount: 51 })],
      ["messages[0].content", randText({ charCount: 4001 })],
      ["messages[0].authorUsername", randText({ charCount: 51 })],
      ["messages[0].authorIdentifier", randText({ charCount: 51 })],
      ["messages[0].authorAvatarUrl", randText({ charCount: 201 })],
      ["title", 123],
      ["appSpecificData", "bad"],
      ["messages[0].appSpecificData", "bad"],
      ["messages[0].authorIdentifier", 123],
      ["messages[0].authorAvatarUrl", 123],
    ])("returns a 404 if %s = %p", async (key, value) => {
      const input: CreateSnippetInput = {
        appId: prisma.app.idToExternalId(1),
        public: true,
        title: "Test snippet title",
        appSpecificData: { key: "value" },
        messages: [
          {
            content: "Content",
            sentAt: new Date(10),
            appSpecificData: { key2: "value2" },
            authorUsername: "Icyspawn",
            authorIdentifier: "1234",
            authorAvatarUrl: "http://example.com/123.png",
          },
        ],
      };
      set(input, key, value);

      await request(app).post("/snippets").send(input).expect(400);
    });

    it("can create a snippet", async () => {
      const appEntity = await prisma.app.create({
        data: {
          name: "Discord",
        },
      });
      await prisma.user.create({
        data: {
          username: "crasken",
          email: "chase.gladish@gmail.com",
        },
      });

      const input: CreateSnippetInput = {
        appId: prisma.app.idToExternalId(appEntity.id),
        public: true,
        title: "Test snippet title",
        appSpecificData: { key: "value" },
        messages: Array(20)
          .fill(null)
          .map((_, index) => ({
            content: `Content ${index}`,
            sentAt: new Date(index * 5),
            appSpecificData: { key: `value ${index}` },
            authorUsername: "Icyspawn",
            authorIdentifier: "1234",
            authorAvatarUrl: "http://example.com/123.png",
          })),
      };

      const response = await request(app)
        .post("/snippets")
        .send(input)
        .expect(201);
      expect(response.body).toMatchSnapshot();
    });
  });
});
