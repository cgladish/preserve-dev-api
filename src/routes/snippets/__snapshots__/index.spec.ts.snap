// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`snippets routes GET /:id can get a snippet by external ID 1`] = `
{
  "app": {
    "id": "zNrpboD",
    "name": "Discord",
  },
  "appSpecificDataJson": "{"key":"value"}",
  "creatorId": "E21K9jm",
  "id": "BQAYLpr",
  "messages": [
    {
      "appSpecificDataJson": "{"key2":"value2"}",
      "authorAvatarUrl": "http://example.com/123.png",
      "authorIdentifier": "1234",
      "authorUsername": "Icyspawn",
      "content": "Content",
      "id": "1GX8zMN",
      "sentAt": "1970-01-01T00:00:00.010Z",
    },
  ],
  "public": true,
  "title": "Test snippet title",
  "views": 0,
}
`;

exports[`snippets routes POST / can create a snippet with auth 1`] = `
{
  "app": {
    "id": "zNrpboD",
    "name": "Discord",
  },
  "appSpecificDataJson": "{"key":"value"}",
  "creatorId": "E21K9jm",
  "id": "BQAYLpr",
  "messages": [
    {
      "appSpecificDataJson": "{"key":"value 0"}",
      "authorAvatarUrl": "http://example.com/123.png",
      "authorIdentifier": "1234",
      "authorUsername": "Icyspawn",
      "content": "Content 0",
      "id": "1GX8zMN",
      "sentAt": "1970-01-01T00:00:00.000Z",
    },
    {
      "appSpecificDataJson": "{"key":"value 1"}",
      "authorAvatarUrl": "http://example.com/123.png",
      "authorIdentifier": "1234",
      "authorUsername": "Icyspawn",
      "content": "Content 1",
      "id": "WKM1Nr1",
      "sentAt": "1970-01-01T00:00:00.005Z",
    },
    {
      "appSpecificDataJson": "{"key":"value 2"}",
      "authorAvatarUrl": "http://example.com/123.png",
      "authorIdentifier": "1234",
      "authorUsername": "Icyspawn",
      "content": "Content 2",
      "id": "E7rEDMv",
      "sentAt": "1970-01-01T00:00:00.010Z",
    },
    {
      "appSpecificDataJson": "{"key":"value 3"}",
      "authorAvatarUrl": "http://example.com/123.png",
      "authorIdentifier": "1234",
      "authorUsername": "Icyspawn",
      "content": "Content 3",
      "id": "pwrLpr8",
      "sentAt": "1970-01-01T00:00:00.015Z",
    },
    {
      "appSpecificDataJson": "{"key":"value 4"}",
      "authorAvatarUrl": "http://example.com/123.png",
      "authorIdentifier": "1234",
      "authorUsername": "Icyspawn",
      "content": "Content 4",
      "id": "R2MVGrV",
      "sentAt": "1970-01-01T00:00:00.020Z",
    },
    {
      "appSpecificDataJson": "{"key":"value 5"}",
      "authorAvatarUrl": "http://example.com/123.png",
      "authorIdentifier": "1234",
      "authorUsername": "Icyspawn",
      "content": "Content 5",
      "id": "GE4lBrZ",
      "sentAt": "1970-01-01T00:00:00.025Z",
    },
    {
      "appSpecificDataJson": "{"key":"value 6"}",
      "authorAvatarUrl": "http://example.com/123.png",
      "authorIdentifier": "1234",
      "authorUsername": "Icyspawn",
      "content": "Content 6",
      "id": "NZXvDMo",
      "sentAt": "1970-01-01T00:00:00.030Z",
    },
    {
      "appSpecificDataJson": "{"key":"value 7"}",
      "authorAvatarUrl": "http://example.com/123.png",
      "authorIdentifier": "1234",
      "authorUsername": "Icyspawn",
      "content": "Content 7",
      "id": "wq4Bz42",
      "sentAt": "1970-01-01T00:00:00.035Z",
    },
    {
      "appSpecificDataJson": "{"key":"value 8"}",
      "authorAvatarUrl": "http://example.com/123.png",
      "authorIdentifier": "1234",
      "authorUsername": "Icyspawn",
      "content": "Content 8",
      "id": "oPrOQ46",
      "sentAt": "1970-01-01T00:00:00.040Z",
    },
    {
      "appSpecificDataJson": "{"key":"value 9"}",
      "authorAvatarUrl": "http://example.com/123.png",
      "authorIdentifier": "1234",
      "authorUsername": "Icyspawn",
      "content": "Content 9",
      "id": "nBXweM2",
      "sentAt": "1970-01-01T00:00:00.045Z",
    },
    {
      "appSpecificDataJson": "{"key":"value 10"}",
      "authorAvatarUrl": "http://example.com/123.png",
      "authorIdentifier": "1234",
      "authorUsername": "Icyspawn",
      "content": "Content 10",
      "id": "gKMxBX8",
      "sentAt": "1970-01-01T00:00:00.050Z",
    },
    {
      "appSpecificDataJson": "{"key":"value 11"}",
      "authorAvatarUrl": "http://example.com/123.png",
      "authorIdentifier": "1234",
      "authorUsername": "Icyspawn",
      "content": "Content 11",
      "id": "EQXQqX7",
      "sentAt": "1970-01-01T00:00:00.055Z",
    },
    {
      "appSpecificDataJson": "{"key":"value 12"}",
      "authorAvatarUrl": "http://example.com/123.png",
      "authorIdentifier": "1234",
      "authorUsername": "Icyspawn",
      "content": "Content 12",
      "id": "K2rka49",
      "sentAt": "1970-01-01T00:00:00.060Z",
    },
    {
      "appSpecificDataJson": "{"key":"value 13"}",
      "authorAvatarUrl": "http://example.com/123.png",
      "authorIdentifier": "1234",
      "authorUsername": "Icyspawn",
      "content": "Content 13",
      "id": "a1X7l43",
      "sentAt": "1970-01-01T00:00:00.065Z",
    },
    {
      "appSpecificDataJson": "{"key":"value 14"}",
      "authorAvatarUrl": "http://example.com/123.png",
      "authorIdentifier": "1234",
      "authorUsername": "Icyspawn",
      "content": "Content 14",
      "id": "gJXn0Me",
      "sentAt": "1970-01-01T00:00:00.070Z",
    },
    {
      "appSpecificDataJson": "{"key":"value 15"}",
      "authorAvatarUrl": "http://example.com/123.png",
      "authorIdentifier": "1234",
      "authorUsername": "Icyspawn",
      "content": "Content 15",
      "id": "6746Z4j",
      "sentAt": "1970-01-01T00:00:00.075Z",
    },
    {
      "appSpecificDataJson": "{"key":"value 16"}",
      "authorAvatarUrl": "http://example.com/123.png",
      "authorIdentifier": "1234",
      "authorUsername": "Icyspawn",
      "content": "Content 16",
      "id": "p845A4a",
      "sentAt": "1970-01-01T00:00:00.080Z",
    },
    {
      "appSpecificDataJson": "{"key":"value 17"}",
      "authorAvatarUrl": "http://example.com/123.png",
      "authorIdentifier": "1234",
      "authorUsername": "Icyspawn",
      "content": "Content 17",
      "id": "Y0rJbrp",
      "sentAt": "1970-01-01T00:00:00.085Z",
    },
    {
      "appSpecificDataJson": "{"key":"value 18"}",
      "authorAvatarUrl": "http://example.com/123.png",
      "authorIdentifier": "1234",
      "authorUsername": "Icyspawn",
      "content": "Content 18",
      "id": "nxMD6Xy",
      "sentAt": "1970-01-01T00:00:00.090Z",
    },
    {
      "appSpecificDataJson": "{"key":"value 19"}",
      "authorAvatarUrl": "http://example.com/123.png",
      "authorIdentifier": "1234",
      "authorUsername": "Icyspawn",
      "content": "Content 19",
      "id": "LR4bLrz",
      "sentAt": "1970-01-01T00:00:00.095Z",
    },
  ],
  "public": true,
  "title": "Test snippet title",
  "views": 0,
}
`;

exports[`snippets routes POST / can create a snippet with empty content message 1`] = `
{
  "app": {
    "id": "zNrpboD",
    "name": "Discord",
  },
  "appSpecificDataJson": "{"key":"value"}",
  "creatorId": "E21K9jm",
  "id": "BQAYLpr",
  "messages": [
    {
      "appSpecificDataJson": "{"key":"value"}",
      "authorAvatarUrl": "http://example.com/123.png",
      "authorIdentifier": "1234",
      "authorUsername": "Icyspawn",
      "content": "",
      "id": "1GX8zMN",
      "sentAt": "1970-01-01T00:00:00.005Z",
    },
  ],
  "public": true,
  "title": "Test snippet title",
  "views": 0,
}
`;

exports[`snippets routes POST / can create a snippet with no auth 1`] = `
{
  "app": {
    "id": "zNrpboD",
    "name": "Discord",
  },
  "appSpecificDataJson": "{"key":"value"}",
  "creatorId": null,
  "id": "BQAYLpr",
  "messages": [
    {
      "appSpecificDataJson": "{"key":"value 0"}",
      "authorAvatarUrl": "http://example.com/123.png",
      "authorIdentifier": "1234",
      "authorUsername": "Icyspawn",
      "content": "Content 0",
      "id": "1GX8zMN",
      "sentAt": "1970-01-01T00:00:00.000Z",
    },
    {
      "appSpecificDataJson": "{"key":"value 1"}",
      "authorAvatarUrl": "http://example.com/123.png",
      "authorIdentifier": "1234",
      "authorUsername": "Icyspawn",
      "content": "Content 1",
      "id": "WKM1Nr1",
      "sentAt": "1970-01-01T00:00:00.005Z",
    },
    {
      "appSpecificDataJson": "{"key":"value 2"}",
      "authorAvatarUrl": "http://example.com/123.png",
      "authorIdentifier": "1234",
      "authorUsername": "Icyspawn",
      "content": "Content 2",
      "id": "E7rEDMv",
      "sentAt": "1970-01-01T00:00:00.010Z",
    },
    {
      "appSpecificDataJson": "{"key":"value 3"}",
      "authorAvatarUrl": "http://example.com/123.png",
      "authorIdentifier": "1234",
      "authorUsername": "Icyspawn",
      "content": "Content 3",
      "id": "pwrLpr8",
      "sentAt": "1970-01-01T00:00:00.015Z",
    },
    {
      "appSpecificDataJson": "{"key":"value 4"}",
      "authorAvatarUrl": "http://example.com/123.png",
      "authorIdentifier": "1234",
      "authorUsername": "Icyspawn",
      "content": "Content 4",
      "id": "R2MVGrV",
      "sentAt": "1970-01-01T00:00:00.020Z",
    },
    {
      "appSpecificDataJson": "{"key":"value 5"}",
      "authorAvatarUrl": "http://example.com/123.png",
      "authorIdentifier": "1234",
      "authorUsername": "Icyspawn",
      "content": "Content 5",
      "id": "GE4lBrZ",
      "sentAt": "1970-01-01T00:00:00.025Z",
    },
    {
      "appSpecificDataJson": "{"key":"value 6"}",
      "authorAvatarUrl": "http://example.com/123.png",
      "authorIdentifier": "1234",
      "authorUsername": "Icyspawn",
      "content": "Content 6",
      "id": "NZXvDMo",
      "sentAt": "1970-01-01T00:00:00.030Z",
    },
    {
      "appSpecificDataJson": "{"key":"value 7"}",
      "authorAvatarUrl": "http://example.com/123.png",
      "authorIdentifier": "1234",
      "authorUsername": "Icyspawn",
      "content": "Content 7",
      "id": "wq4Bz42",
      "sentAt": "1970-01-01T00:00:00.035Z",
    },
    {
      "appSpecificDataJson": "{"key":"value 8"}",
      "authorAvatarUrl": "http://example.com/123.png",
      "authorIdentifier": "1234",
      "authorUsername": "Icyspawn",
      "content": "Content 8",
      "id": "oPrOQ46",
      "sentAt": "1970-01-01T00:00:00.040Z",
    },
    {
      "appSpecificDataJson": "{"key":"value 9"}",
      "authorAvatarUrl": "http://example.com/123.png",
      "authorIdentifier": "1234",
      "authorUsername": "Icyspawn",
      "content": "Content 9",
      "id": "nBXweM2",
      "sentAt": "1970-01-01T00:00:00.045Z",
    },
    {
      "appSpecificDataJson": "{"key":"value 10"}",
      "authorAvatarUrl": "http://example.com/123.png",
      "authorIdentifier": "1234",
      "authorUsername": "Icyspawn",
      "content": "Content 10",
      "id": "gKMxBX8",
      "sentAt": "1970-01-01T00:00:00.050Z",
    },
    {
      "appSpecificDataJson": "{"key":"value 11"}",
      "authorAvatarUrl": "http://example.com/123.png",
      "authorIdentifier": "1234",
      "authorUsername": "Icyspawn",
      "content": "Content 11",
      "id": "EQXQqX7",
      "sentAt": "1970-01-01T00:00:00.055Z",
    },
    {
      "appSpecificDataJson": "{"key":"value 12"}",
      "authorAvatarUrl": "http://example.com/123.png",
      "authorIdentifier": "1234",
      "authorUsername": "Icyspawn",
      "content": "Content 12",
      "id": "K2rka49",
      "sentAt": "1970-01-01T00:00:00.060Z",
    },
    {
      "appSpecificDataJson": "{"key":"value 13"}",
      "authorAvatarUrl": "http://example.com/123.png",
      "authorIdentifier": "1234",
      "authorUsername": "Icyspawn",
      "content": "Content 13",
      "id": "a1X7l43",
      "sentAt": "1970-01-01T00:00:00.065Z",
    },
    {
      "appSpecificDataJson": "{"key":"value 14"}",
      "authorAvatarUrl": "http://example.com/123.png",
      "authorIdentifier": "1234",
      "authorUsername": "Icyspawn",
      "content": "Content 14",
      "id": "gJXn0Me",
      "sentAt": "1970-01-01T00:00:00.070Z",
    },
    {
      "appSpecificDataJson": "{"key":"value 15"}",
      "authorAvatarUrl": "http://example.com/123.png",
      "authorIdentifier": "1234",
      "authorUsername": "Icyspawn",
      "content": "Content 15",
      "id": "6746Z4j",
      "sentAt": "1970-01-01T00:00:00.075Z",
    },
    {
      "appSpecificDataJson": "{"key":"value 16"}",
      "authorAvatarUrl": "http://example.com/123.png",
      "authorIdentifier": "1234",
      "authorUsername": "Icyspawn",
      "content": "Content 16",
      "id": "p845A4a",
      "sentAt": "1970-01-01T00:00:00.080Z",
    },
    {
      "appSpecificDataJson": "{"key":"value 17"}",
      "authorAvatarUrl": "http://example.com/123.png",
      "authorIdentifier": "1234",
      "authorUsername": "Icyspawn",
      "content": "Content 17",
      "id": "Y0rJbrp",
      "sentAt": "1970-01-01T00:00:00.085Z",
    },
    {
      "appSpecificDataJson": "{"key":"value 18"}",
      "authorAvatarUrl": "http://example.com/123.png",
      "authorIdentifier": "1234",
      "authorUsername": "Icyspawn",
      "content": "Content 18",
      "id": "nxMD6Xy",
      "sentAt": "1970-01-01T00:00:00.090Z",
    },
    {
      "appSpecificDataJson": "{"key":"value 19"}",
      "authorAvatarUrl": "http://example.com/123.png",
      "authorIdentifier": "1234",
      "authorUsername": "Icyspawn",
      "content": "Content 19",
      "id": "LR4bLrz",
      "sentAt": "1970-01-01T00:00:00.095Z",
    },
  ],
  "public": true,
  "title": "Test snippet title",
  "views": 0,
}
`;

exports[`snippets routes POST / can create a snippet with no auth 2`] = `
{
  "appId": 1,
  "appSpecificDataJson": "{"key":"value"}",
  "creatorId": null,
  "id": 1,
  "public": true,
  "title": "Test snippet title",
  "views": 0,
}
`;
